doc/source/ray-observability/user-guides/debug-apps/debug-memory.rst

```diff
@@ -1,8 +1,182 @@
 .. _ray-core-mem-profiling:
 
 Debugging Memory Issues
+=======================
+
+
+.. _troubleshooting-out-of-memory:
+
+Debugging Out of Memory
 -----------------------
 
+Before reading this section, it is recommended to understand Ray's :ref:`Memory Management <memory>` model.
+
+- To find if your cluster has out-of-memory problems, view :ref:`How to Detect Out-of-Memory Errors? <troubleshooting-out-of-memory-how-to-detect>`.
+- To find what leaks the memory, view :ref:`Find per Task and Actor Memory Usage <troubleshooting-out-of-memory-task-actor-mem-usage>`.
+- If your head node has high memory usage, view :ref:`Head Node Out-of-Memory Error <troubleshooting-out-of-memory-head>`.
+- If your memory usage is high due to high parallelism, view :ref:`Reduce Parallelism <troubleshooting-out-of-memory-reduce-parallelism>`.
+- If you want to profile per task and actor memory usage, view :ref:`Profile Task and Actor Memory Usage <troubleshooting-out-of-memory-profile>`.
+
+What's the Out-of-Memory Error?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Memory is a limited resource. When a process requests memory and the OS fails to allocate memory, the OS executes a routine to free up memory
+by killing a process that has high memory usage (via SIGKILL) to avoid the OS becoming unstable. It is called `Linux Out of Memory killer <https://www.kernel.org/doc/gorman/html/understand/understand016.html>`_.
+
+One of the common problems of the Linux out-of-memory killer is that processes are killed by a SIGKILL without Ray noticing it. 
+Since SIGKILL cannot be handled by processes, it makes Ray difficult to raise a proper error message
+and take proper actions for fault tolerance.
+To solve this problem, Ray built and enabled (from Ray 2.2) an application-level :ref:`memory monitor <ray-oom-monitor>`,
+which keeps monitoring the memory usage of the host and kills the Ray workers before the Linux out-of-memory killer kicks in. 
+
+.. _troubleshooting-out-of-memory-how-to-detect:
+
+How to Detect Out-of-Memory Errors?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If tasks or actors are killed by the Linux out-of-memory killer, Ray worker processes are unable to catch and display an exact root cause
+because SIGKILL cannot be handled by processes. If you call ``ray.get`` into the tasks and actors that were executed from the dead worker,
```

Comment:
```suggestion
because SIGKILL cannot be handled by processes. If you call ``ray.get`` into the Tasks and Actors that were executed from the dead worker,
```